polygeist/include/mlir/Dialect/Polygeist/Utils/TransformUtils.h

```diff
@@ -230,6 +237,66 @@ class LoopTools {
                    const LoopVersionCondition &versionCond) const;
 };
 
+//===----------------------------------------------------------------------===//
+// VersionConditionBuilder
+//===----------------------------------------------------------------------===//
+
+using AccessorType = TypedValue<MemRefType>;
+using AccessorPairType = std::pair<AccessorType, AccessorType>;
+class VersionConditionBuilder {
+public:
+  VersionConditionBuilder(
+      LoopLikeOpInterface loop,
+      ArrayRef<AccessorPairType> requireNoOverlapAccessorPairs)
+      : loop(loop), accessorPairs(requireNoOverlapAccessorPairs) {}
+
+  using SCFCondition = LoopVersionCondition::SCFCondition;
+  using AffineCondition = LoopVersionCondition::AffineCondition;
+
+  std::unique_ptr<LoopVersionCondition> createCondition() const {
+    OpBuilder builder(loop);
+    Location loc = loop.getLoc();
+    SCFCondition scfCond = createSCFCondition(builder, loc);
+    return std::make_unique<LoopVersionCondition>(scfCond);
+  }
+
+private:
+  /// Create a versioning condition suitable for scf::IfOp.
+  SCFCondition createSCFCondition(OpBuilder builder, Location loc) const;
+
+  template <typename OpTy>
```

Comment:
Given the class is now in a header file I would place this static functions in the .cpp file to keep the interface smaller and easier to read.